File: models/todo.js
const mongoose = require("mongoose");

const ObjectId = mongoose.Schema.Types.ObjectId;

const todoSchema = new mongoose.Schema({
    projectId: { type: ObjectId, ref: "Project" },
    name: { type: String, required: true },
    description: { type: String, default: "" },
    state: { type: String, default: "todo" },
    priority: { type: String, default: "low" },
    tags: [{ type: String }],
    dueDate: {
        startAt: { type: Date, default: Date.now },
        endAt: { type: Date, default: null },
    },
    isDone: { type: Boolean, default: false },
    isDeleted: { type: Boolean, default: false },
    isArchived: { type: Boolean, default: false },
    isPinned: { type: Boolean, default: false },
    createdAt: { type: Date, default: Date.now },
    updatedAt: { type: Date, default: Date.now },
});

const Todo = mongoose.model("Todo", todoSchema);

module.exports = Todo;

File: models/todo.go
package models

import "time"

type Todo struct {
	Model

	UserId       int64  `gorm:"not null" json:"userId"`
	ProjectId    int64  `gorm:"not null" json:"projectId"`
	ParentTodoId *int64 `gorm:"null" json:"parentTodoId"`

	Name        string     `gorm:"size:64" json:"name"`
	Description string     `gorm:"size:256" json:"description"`
	State       int8       `json:"state"`
	Priority    int8       `json:"priority"`
	StartAt     *time.Time `gorm:"null" json:"startAt"`
	EndAt       *time.Time `gorm:"null" json:"endAt"`
	ArchivedAt  *time.Time `gorm:"null" json:"archivedAt"`
	FavoritedAt *time.Time `json:"favoritedAt"`
	GivenUpAt   *time.Time `json:"givenUpAt"`
	Tags        []string   `gorm:"serializer:json;type:json" json:"tags"`
}

type TodoResponse struct {
	ModelResponse

	UserId       string  `json:"userId"`
	ProjectId    string  `json:"projectId"`
	ParentTodoId *string `json:"parentTodoId"`

	Name        string     `json:"name"`
	Description string     `json:"description"`
	State       string     `json:"state"`
	Priority    string     `json:"priority"`
	StartAt     *time.Time `json:"startAt"`
	EndAt       *time.Time `json:"endAt"`
	ArchivedAt  *time.Time `json:"archivedAt"`
	FavoritedAt *time.Time `json:"favoritedAt"`
	GivenUpAt   *time.Time `json:"givenUpAt"`
	Tags        []string   `json:"tags"`

	// extras fill
	IsArchived  bool `json:"isArchived"`
	IsDeleted   bool `json:"isDeleted"`
	IsFavorited bool `json:"isFavorited"`
	IsGivenUp   bool `json:"isGivenUp"`
}
